### PR TITLE
[reflection] Only duplicate MonoMarshalSpec strings for custom types

### LINKED ISSUE #16570: Crash on GetParameters() with UnmanagedType.LPArray in dynamic assembly
```csharp
using System;
using System.Reflection;
using System.Reflection.Emit;
using System.Runtime.InteropServices;

public class Program {
	public static void Main ()
	{
		var aName = new AssemblyName("TestAssembly");
		var testAssembly = AppDomain.CurrentDomain.DefineDynamicAssembly(aName, AssemblyBuilderAccess.RunAndSave);
		var testModule = testAssembly.DefineDynamicModule(aName.Name, aName.Name + ".dll");

		var typeBuilder = testModule.DefineType("TestType");

		var ctorBuilder = typeBuilder.DefineConstructor(MethodAttributes.Public, CallingConventions.Standard, Type.EmptyTypes);

		var ctorIl = ctorBuilder.GetILGenerator();
		ctorIl.Emit(OpCodes.Ret);

		var methodBuilder = typeBuilder.DefineMethod("TestMethod", MethodAttributes.Public, typeof(void), new[] { typeof(int[]) });
		methodBuilder.DefineParameter(0, ParameterAttributes.Retval, null);
		var paramBuilder = methodBuilder.DefineParameter(1, ParameterAttributes.None, null);

		var attrCtor = typeof(MarshalAsAttribute).GetConstructor(new[] { typeof(UnmanagedType) });
		object[] ctorArgs = { UnmanagedType.LPArray }; // every other value works fine
		var attr = new CustomAttributeBuilder(attrCtor, ctorArgs);
		paramBuilder.SetCustomAttribute(attr);

		var methodIl = methodBuilder.GetILGenerator();
		methodIl.Emit(OpCodes.Ret);

		var createdType = typeBuilder.CreateType();

		var methodInfo = createdType.GetMethod("TestMethod", BindingFlags.Instance | BindingFlags.Public);
		methodInfo.GetParameters(); // crash!

		Console.WriteLine("Done"); // not going here
		Console.ReadKey();
	}
}

```

### Current Behavior
The most interesting part is the crash only occures when I use `UnmanagedType.LPArray` in `MarshalAs`. I've tried most other values from `UnmanagedType` and their work fine in this example (even though the parameter is `int[]`, actually, the type of parameter doesn't seem to matter here).

This looks similar to #10201 but repro is different.

## On which platforms did you notice this

[ ] macOS
[ ] Linux
[X] Windows

**Version Used**:
```
Mono JIT compiler version 5.18.0 (Visual Studio built mono)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       normal
        Notification:  Thread + polling
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        Suspend:       preemptive
        GC:            sgen (concurrent by default)
```
## Stacktrace
(taken from Unity crash log)
```
(mono-2.0-bdwgc) mono_method_get_marshal_info
(mono-2.0-bdwgc) mono_type_get_object
(mono-2.0-bdwgc) mono_verify_corlib
(mono-2.0-bdwgc) mono_param_get_objects
(mscorlib) System.Reflection.MonoMethodInfo.get_parameter_info()
(mscorlib) System.Reflection.MonoMethodInfo.GetParametersInfo()
(mscorlib) System.Reflection.MonoMethod.GetParameters()
```

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Fixes #16570

We were treating the union as though it was a different type than it really is, and so we got junk data for the pointers and attempting to access them caused a crash.

Also fixed a typo seen along the way.

### CODE REVIEW

File: mcs/class/corlib/Test/System.Reflection.Emit/CustomAttributeBuilderTest.cs [hunk 1]
```diff
@@ -899,6 +899,40 @@ private static void AddCustomClassAttribute (TypeBuilder typeBuilder, Type custo
...
+			paramBuilder.SetCustomAttribute(attr);
+
+			var methodIl = methodBuilder.GetILGenerator();
+			methodIl.Emit(OpCodes.Ret);
+
+			var createdType = typeBuilder.CreateType();
+
+			var methodInfo = createdType.GetMethod("TestMethod", BindingFlags.Instance | BindingFlags.Public);
+			methodInfo.GetParameters();
+		}
```
Comment (lambdageek) [inline mcs/class/corlib/Test/System.Reflection.Emit/CustomAttributeBuilderTest.cs]: You should probably assert something in the test.

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build API Diff

**Comment by CoffeeFlux** [comment]
@monojenkins build Project Files Diff

**Comment by lambdageek** [comment]
@monojenkins backport to 2019-08

**Pull request merged**